### ISSUE TITLE
Optimistic lock does not work correctly with REST call

### ISSUE THREAD

**Issue by Dennisber00** [spring-projects/spring-framework#26365]
**Affects:**  Spring boot 2.4.1 and earlier
**Affects:**  spring-boot-starter-data-jpa 2.4.1 and earlier
**Affects:**  spring-boot-starter-web 2.4.1 and earlier

**Description:** I found out that optimistic lock does not work correctly in a case when a method, which can produce it, is called via REST controller.

As an example, the following method of a service must produce an optimistic lock exception every time it is called (let assume that a book entity is already created in the database):

```java
1  public void produceOptimisticLock(Long id) {
2   Book found = bookRepository.findById(id)
3        .orElseThrow(() -> new RuntimeException("Oops"));
4    found.setName("optLock !");
5    found.setVersion(-777L);
6    save(found);
7  }
```

Because there is no _@Transaction_ annotation above the method and a read only transaction as well as hibernate session are created on line 2, then entity is loaded and got a DETACHED status. After that on line 6 a new transaction as well as hibernate session are created to merge changes in "found" entity. Finally _DefaultMergeEventListener_ (hibernate) produces a new optimistic lock exception, because the entity is processed as DETACHED (_entityIsDetached_ method) and the version field was changed. It works correctly when it is called internally in an application. 

But it does not work at all when it is called as part of REST request handling. Because _OpenEntityManagerInViewFilter_ delegates _OpenEntityManagerInViewInterceptor_ to create a new entity manager instance which creates one hibernate session to process entire request. In this case on line 2 the "found" entity is cached by _StatefulPersistenceContext_ which leads to a problem on line 6, because the session is still alive/active _StatefulPersistenceContext_ still contains "found" entity in L1 cache. In this case entity state is resolved as PERSISTENT in _DefaultMergeEventListener.onMerge_ method. It means that "found" entity is processed as  PERSISTENT in _entityIsPersistent_ method of _DefaultMergeEventListener_. _entityIsPersistent_ does not check if the version is changed. Because of it "found" entity is saved on line 6 without any optimistic lock exception.

I guess this method should work always identical and produces the optimistic lock exception. Am i right ? 

I created a demo project to produce this problem:
https://github.com/Dennisber00/spring-opt-log-bug

**Step to reproduce in the demo project**

No optimistic lock exception:
1. Run the app
2. Send a rest call -  http://localhost:8080/book/opt-lock/1
3. Open h2 console http://localhost:8088/h2-console/
4. Execute "select * from book"
5. Check that entity with ID - 1 was successfully updated (version was increased to 1)

Optimistic lock exception:
1. Open application.properties
2. Change _do.magic.fire.opt.lock=false_ -> _do.magic.fire.opt.lock=true_ property
3. Run the app
4. The app should be crashed with optimistic lock exception due to _DoMagic_ event listener class which calls _produceOptimisticLock_ method.
---

**Comment by Dennisber00** [comment]
did i create the issue incorrectly ? or in wrong place ?

**Comment by wwwrakesh1** [comment]
I have seen same issue

**Comment by snicoll** [comment]
@Dennisber00 this all looks like expected, setting the version manually and not having a transaction around that code should throw that exception. I am going to close this now and please use StackOverflow if you have more questions around JPA usage.

**Comment by Dennisber00** [comment]
> @Dennisber00 this all looks like expected, setting the version manually and not having a transaction around that code should throw that exception. I am going to close this now and please use StackOverflow if you have more questions around JPA usage.

@snicoll Have you read my description carefully ? Of course, it is excepted to have en exception, but "But it does not work at all when it is called as part of REST request handling." It means no exception is thrown in this case ! As you said, exception is expected here, but no exception means it is a bug. Please, re-read my description one more time.

**Comment by snicoll** [comment]
> I guess this method should work always identical and produces the optimistic lock exception. Am i right ?"

No. I've shared why I think that code is invalid already. If you read an entity outside of a transaction and attempt to save them, you're opening yourself to inconsistencies like this. You think think this is a Spring issue but what you use from it simply delegates to JPA, that's why I mentioned to ask usage question on SO.

**Comment by Dennisber00** [comment]
> No. I've shared why I think that code is invalid already. If you read an entity outside of a transaction and attempt to save it the way you do, you're opening yourself to inconsistencies like this. You think this is a Spring issue but what you use from it simply delegates to JPA, that's why I mentioned to ask usage question on SO.

@snicoll This is only a simplified sample (invalid as you wish) to demonstrate this problem. Of course, in this sample it is only JPA delegation. You are right here. Unfortunately, i am not allow to share an Enterprise project (which is definitely way more complex than this sample) to show a real example where this problem was faced.

No exception is thrown because of OSIV or OpenSessionInViewFilter creating a session per each REST request (spring logic and not related to JPA) + lack of transaction (human error). And **OSIV is enabled by default**. It means that in a real big project where spring/spring boot is used to simplify developers life ("easy to use" is one of spring slogan, right ?) if human factor is applied (or i would say when applied) and @Transaction is lost somewhere in between (this situation is easy to imagine, unfortunately) a developer will face inconsistency in behavior (as you correctly mentioned) which is really hard to identify in a spot where exception must be thrown. Based on this i think it is still a spring issue, not JPA or something else.
